File: src/cee/engine/t3_geodesic.py
"""
T3 — Project Geodesic: 测地线导航引擎

核心思想: 将代码生成/文本生成转化为高维流形上的测地线导航。
在多个可能解之间找到最优路径(Geodesic)。

证据: 3条路径探索，2条新颖路径，原子层破局
"""

import ast
import textwrap
from dataclasses import dataclass, field
from typing import Any

import numpy as np


@dataclass
class CodeLandmark:
    """代码流形上的路标点."""
    position: list[float]       # 流形坐标
    code_snippet: str           # 对应的代码
    quality_score: float        # 局部质量评分
    description: str = ""       # 路标描述


@dataclass
class GeodesicPath:
    """测地线路径."""
    landmarks: list[CodeLandmark]
    total_distance: float       # 总测地线距离
    smoothness: float          # 路径平滑度
    novelty_score: float       # 新颖性分数
    label: str = ""            # 路径标签


class GeodesicNavigationEngine:
    """T3 测地线导航引擎"""

    def __init__(self, n_paths: int = 3, exploration_factor: float = 0.3):
        self.n_paths = n_paths
        self.exploration_factor = exploration_factor

    def extract_landmarks(self, code: str) -> list[CodeLandmark]:
        """
        从代码中提取路标点。

        将代码解析为AST，将每个关键节点映射为流形上的点。
        """
        try:
            tree = ast.parse(code)
        except SyntaxError:
            return self._fallback_landmarks(code)

        landmarks = []
        for node in ast.walk(tree):
            if isinstance(node, (ast.FunctionDef, ast.ClassDef, ast.AsyncFunctionDef)):
                pos = self._node_to_position(node, landmarks)
                snippet = ast.unparse(node) if hasattr(ast, 'unparse') else str(node)
                landmarks.append(CodeLandmark(
                    position=pos,
                    code_snippet=snippet,
                    quality_score=self._local_quality(node),
                    description=self._describe_node(node),
                ))

        if not landmarks:
            landmarks = self._fallback_landmarks(code)

        return landmarks

    def _fallback_landmarks(self, code: str) -> list[CodeLandmark]:
        lines = [l for l in code.split("\n") if l.strip()]
        n = len(lines)
        if n == 0:
            return []

        n_landmarks = min(5, n)
        indices = np.linspace(0, n - 1, n_landmarks, dtype=int)
        landmarks = []
        for i, idx in enumerate(indices):
            landmarks.append(CodeLandmark(
                position=[i / n_landmarks, 0.5, 0.0],
                code_snippet=lines[idx].strip(),
                quality_score=0.5,
                description=f"line-{idx + 1}",
            ))
        return landmarks

    def _node_to_position(self, node: ast.AST,
                           existing: list[CodeLandmark]) -> list[float]:
        base = len(existing)
        col = node.col_offset if hasattr(node, 'col_offset') else 0
        return [
            base / max(base + 1, 1),
            min(col / 80.0, 1.0),
            np.random.uniform(-0.1, 0.1),  # 小随机扰动
        ]

    def _local_quality(self, node: ast.AST) -> float:
        score = 0.6
        if isinstance(node, ast.FunctionDef):
            if node.returns:
                score += 0.1
            if len(node.args.args) <= 5:
                score += 0.1
            if ast.get_docstring(node):
                score += 0.1
        elif isinstance(node, ast.ClassDef):
            if len(node.bases) > 0:
                score += 0.15
            if ast.get_docstring(node):
                score += 0.1
        return min(1.0, score)

    def _describe_node(self, node: ast.AST) -> str:
        if isinstance(node, ast.FunctionDef):
            return f"function:{node.name}"
        elif isinstance(node, ast.ClassDef):
            return f"class:{node.name}"
        elif isinstance(node, ast.AsyncFunctionDef):
            return f"async:{node.name}"
        return f"node:{type(node).__name__}"

    def find_geodesic_paths(self, code: str) -> list[GeodesicPath]:
        """
        在代码流形上寻找多条测地线路径。

        每条路径代表一种实现方案或代码结构优化路径。
        """
        landmarks = self.extract_landmarks(code)
        if len(landmarks) < 2:
            return [GeodesicPath(
                landmarks=landmarks,
                total_distance=0.0,
                smoothness=1.0,
                novelty_score=0.0,
                label="trivial",
            )]

        paths = []
        for p in range(self.n_paths):
            if p == 0:
                ordered = sorted(landmarks, key=lambda l: l.position[0])
                label = "natural-order"
            elif p == 1:
                ordered = sorted(landmarks, key=lambda l: l.quality_score, reverse=True)
                label = "quality-priority"
            elif p == 2:
                ordered = sorted(landmarks, key=lambda l: l.position[1], reverse=True)
                label = "structural-alternative"
            else:
                rng = np.random.RandomState(p)
                indices = list(range(len(landmarks)))
                rng.shuffle(indices)
                ordered = [landmarks[i] for i in indices]
                label = f"exploratory-{p}"

            distance = self._geodesic_distance(ordered)
            smoothness = self._path_smoothness(ordered)
            novelty = self._compute_novelty(ordered, [l for l in landmarks if l not in ordered])

            paths.append(GeodesicPath(
                landmarks=ordered,
                total_distance=distance,
                smoothness=smoothness,
                novelty_score=novelty,
                label=label,
            ))

        return paths

    def _geodesic_distance(self, landmarks: list[CodeLandmark]) -> float:
        if len(landmarks) < 2:
            return 0.0
        total = 0.0
        for i in range(len(landmarks) - 1):
            p1 = np.array(landmarks[i].position)
            p2 = np.array(landmarks[i + 1].position)
            total += np.linalg.norm(p2 - p1)
        return float(total)

    def _path_smoothness(self, landmarks: list[CodeLandmark]) -> float:
        if len(landmarks) < 3:
            return 1.0
        angles = []
        for i in range(1, len(landmarks) - 1):
            v1 = np.array(landmarks[i].position) - np.array(landmarks[i - 1].position)
            v2 = np.array(landmarks[i + 1].position) - np.array(landmarks[i].position)
            n1, n2 = np.linalg.norm(v1), np.linalg.norm(v2)
            if n1 > 0 and n2 > 0:
                cos_angle = np.clip(np.dot(v1, v2) / (n1 * n2), -1, 1)
                angles.append(np.arccos(cos_angle))
        if not angles:
            return 1.0
        avg_angle = np.mean(angles)
        return float(np.exp(-avg_angle / np.pi))

    def _compute_novelty(self, ordered: list[CodeLandmark],
                           excluded: list[CodeLandmark]) -> float:
        if not excluded:
            return 0.0
        ordered_quality = np.mean([l.quality_score for l in ordered])
        excluded_quality = np.mean([l.quality_score for l in excluded])
        novelty = abs(ordered_quality - excluded_quality)
        return min(1.0, novelty / 0.3)

    def recommend_path(self, paths: list[GeodesicPath],
                        prefer_novelty: bool = True) -> GeodesicPath:
        """推荐最优路径(综合质量、平滑度、新颖性)."""
        if not paths:
            raise ValueError("No paths available")

        best = paths[0]
        best_score = -1.0
        for path in paths:
            avg_quality = np.mean([l.quality_score for l in path.landmarks]) if path.landmarks else 0.0
            score = (
                0.4 * avg_quality
                + 0.3 * path.smoothness
                + 0.3 * (path.novelty_score if prefer_novelty else (1 - path.novelty_score))
            )
            if score > best_score:
                best_score = score
                best = path
        return best

    def generate_implementation_template(self, path: GeodesicPath) -> str:
        """基于测地线路径生成实现模板."""
        if not path.landmarks:
            return "# No implementation landmarks found"

        lines = ["# Generated from geodesic path: " + path.label]
        lines.append(f"# Distance: {path.total_distance:.2f}, Smoothness: {path.smoothness:.2f}")
        lines.append("")

        for lm in path.landmarks:
            lines.append(f"# [{lm.description}] quality={lm.quality_score:.2f}")

        return "\n".join(lines)

File: middle-end/src/cee/engine/t3_geodesic.ts
/**
 * T3 — Project Geodesic: 测地线导航引擎
 *
 * 核心思想: 将代码生成/文本生成转化为高维流形上的测地线导航。
 */

function _genId(): string {
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}

class SeededRandom {
  private seed: number;

  constructor(seed: number) {
    this.seed = seed >>> 0;
  }

  next(): number {
    this.seed = (this.seed * 1664525 + 1013904223) >>> 0;
    return this.seed / 4294967296;
  }

  shuffle<T>(array: T[]): T[] {
    const arr = array.slice();
    for (let i = arr.length - 1; i > 0; i--) {
      const j = Math.floor(this.next() * (i + 1));
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    return arr;
  }
}

function mean(values: number[]): number {
  if (!values.length) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function l2Norm(v: number[]): number {
  return Math.sqrt(v.reduce((sum, x) => sum + x * x, 0));
}

function dot(a: number[], b: number[]): number {
  return a.reduce((sum, x, i) => sum + x * b[i], 0);
}

export interface CodeLandmark {
  position: number[];
  code_snippet: string;
  quality_score: number;
  description: string;
}

export interface GeodesicPath {
  landmarks: CodeLandmark[];
  total_distance: number;
  smoothness: number;
  novelty_score: number;
  label: string;
}

export class GeodesicNavigationEngine {
  n_paths: number;
  exploration_factor: number;

  constructor(n_paths = 3, exploration_factor = 0.3) {
    this.n_paths = n_paths;
    this.exploration_factor = exploration_factor;
  }

  private extract_landmarks(code: string): CodeLandmark[] {
    const landmarks = this.parseCodeLandmarks(code);
    if (landmarks.length) return landmarks;
    return this.fallback_landmarks(code);
  }

  private parseCodeLandmarks(code: string): CodeLandmark[] {
    const landmarks: CodeLandmark[] = [];
    const lines = code.split("\n");

    const patterns = [
      { regex: /function\s+(\w+)\s*\(/, type: "function" },
      { regex: /async\s+function\s+(\w+)\s*\(/, type: "async" },
      { regex: /class\s+(\w+)\b/, type: "class" },
      { regex: /const\s+(\w+)\s*=\s*(?:async\s*)?\([^)]*\)\s*=>/, type: "arrow" },
      { regex: /(?:public|private|protected)?\s*(?:async\s*)?(\w+)\s*\([^)]*\)\s*[:{]/, type: "method" },
    ];

    for (let lineIdx = 0; lineIdx < lines.length; lineIdx++) {
      const line = lines[lineIdx];
      for (const { regex, type } of patterns) {
        const match = regex.exec(line);
        if (!match) continue;
        const name = match[1];
        const col = match.index;
        const pos = this.node_to_position(lineIdx, col, landmarks.length);
        const snippet = line.trim();
        landmarks.push({
          position: pos,
          code_snippet: snippet,
          quality_score: this.local_quality(type, line),
          description: `${type}:${name}`,
        });
      }
    }

    return landmarks;
  }

  private fallback_landmarks(code: string): CodeLandmark[] {
    const lines = code.split("\n").filter(l => l.trim());
    const n = lines.length;
    if (!n) return [];

    const nLandmarks = Math.min(5, n);
    const landmarks: CodeLandmark[] = [];
    for (let i = 0; i < nLandmarks; i++) {
      const idx = Math.round((i / (nLandmarks - 1)) * (n - 1));
      landmarks.push({
        position: [i / nLandmarks, 0.5, 0.0],
        code_snippet: lines[idx].trim(),
        quality_score: 0.5,
        description: `line-${idx + 1}`,
      });
    }
    return landmarks;
  }

  private node_to_position(line: number, col: number, existingCount: number): number[] {
    const base = existingCount;
    return [
      base / Math.max(base + 1, 1),
      Math.min(col / 80.0, 1.0),
      Math.random() * 0.2 - 0.1,
    ];
  }

  private local_quality(type: string, line: string): number {
    let score = 0.6;
    if (type === "function" || type === "async" || type === "arrow") {
      if (/\)\s*:\s*\w+/.test(line)) score += 0.1;
      if (line.split(",").length <= 5) score += 0.1;
      if (/\/\*\*|\/\//.test(line)) score += 0.1;
    } else if (type === "class") {
      if (/extends|implements/.test(line)) score += 0.15;
      if (/\/\*\*/.test(line)) score += 0.1;
    }
    return Math.min(1.0, score);
  }

  find_geodesic_paths(code: string): GeodesicPath[] {
    const landmarks = this.extract_landmarks(code);
    if (landmarks.length < 2) {
      return [{
        landmarks,
        total_distance: 0.0,
        smoothness: 1.0,
        novelty_score: 0.0,
        label: "trivial",
      }];
    }

    const paths: GeodesicPath[] = [];
    for (let p = 0; p < this.n_paths; p++) {
      let ordered: CodeLandmark[];
      let label: string;

      if (p === 0) {
        ordered = landmarks.slice().sort((a, b) => a.position[0] - b.position[0]);
        label = "natural-order";
      } else if (p === 1) {
        ordered = landmarks.slice().sort((a, b) => b.quality_score - a.quality_score);
        label = "quality-priority";
      } else if (p === 2) {
        ordered = landmarks.slice().sort((a, b) => b.position[1] - a.position[1]);
        label = "structural-alternative";
      } else {
        const rng = new SeededRandom(p);
        const indices = rng.shuffle(landmarks.map((_, i) => i));
        ordered = indices.map(i => landmarks[i]);
        label = `exploratory-${p}`;
      }

      const distance = this.geodesic_distance(ordered);
      const smoothness = this.path_smoothness(ordered);
      const novelty = this.compute_novelty(ordered, landmarks.filter(l => !ordered.includes(l)));

      paths.push({
        landmarks: ordered,
        total_distance: distance,
        smoothness,
        novelty_score: novelty,
        label,
      });
    }

    return paths;
  }

  private geodesic_distance(landmarks: CodeLandmark[]): number {
    if (landmarks.length < 2) return 0.0;
    let total = 0.0;
    for (let i = 0; i < landmarks.length - 1; i++) {
      const p1 = landmarks[i].position;
      const p2 = landmarks[i + 1].position;
      const diff = p2.map((v, j) => v - p1[j]);
      total += l2Norm(diff);
    }
    return total;
  }

  private path_smoothness(landmarks: CodeLandmark[]): number {
    if (landmarks.length < 3) return 1.0;
    const angles: number[] = [];
    for (let i = 1; i < landmarks.length - 1; i++) {
      const v1 = landmarks[i].position.map((v, j) => v - landmarks[i - 1].position[j]);
      const v2 = landmarks[i + 1].position.map((v, j) => v - landmarks[i].position[j]);
      const n1 = l2Norm(v1);
      const n2 = l2Norm(v2);
      if (n1 > 0 && n2 > 0) {
        const cosAngle = Math.max(-1, Math.min(1, dot(v1, v2) / (n1 * n2)));
        angles.push(Math.acos(cosAngle));
      }
    }
    if (!angles.length) return 1.0;
    const avgAngle = mean(angles);
    return Math.exp(-avgAngle / Math.PI);
  }

  private compute_novelty(ordered: CodeLandmark[], excluded: CodeLandmark[]): number {
    if (!excluded.length) return 0.0;
    const orderedQuality = mean(ordered.map(l => l.quality_score));
    const excludedQuality = mean(excluded.map(l => l.quality_score));
    const novelty = Math.abs(orderedQuality - excludedQuality);
    return Math.min(1.0, novelty / 0.3);
  }

  recommend_path(paths: GeodesicPath[], prefer_novelty = true): GeodesicPath {
    if (!paths.length) {
      throw new Error("No paths available");
    }

    let best = paths[0];
    let bestScore = -1.0;
    for (const path of paths) {
      const avgQuality = path.landmarks.length ? mean(path.landmarks.map(l => l.quality_score)) : 0.0;
      const score = (
        0.4 * avgQuality
        + 0.3 * path.smoothness
        + 0.3 * (prefer_novelty ? path.novelty_score : (1 - path.novelty_score))
      );
      if (score > bestScore) {
        bestScore = score;
        best = path;
      }
    }
    return best;
  }

  generate_implementation_template(path: GeodesicPath): string {
    if (!path.landmarks.length) return "# No implementation landmarks found";
    const lines: string[] = [
      `# Generated from geodesic path: ${path.label}`,
      `# Distance: ${path.total_distance.toFixed(2)}, Smoothness: ${path.smoothness.toFixed(2)}`,
      "",
    ];
    for (const lm of path.landmarks) {
      lines.push(`# [${lm.description}] quality=${lm.quality_score.toFixed(2)}`);
    }
    return lines.join("\n");
  }
}

export { _genId };
